controller/konnect/ops_kongconsumer.go

```diff
@@ -0,0 +1,195 @@
+			metav1.ConditionTrue,
+			KonnectEntityProgrammedReason,
+			"",
+			consumer.GetGeneration(),
+		),
+		consumer,
+	)
+
+	return nil
+}
+
+func updateConsumer(
+	ctx context.Context,
+	sdk *sdkkonnectgo.SDK,
+	cl client.Client,
+	consumer *configurationv1.KongConsumer,
+) error {
+	// TODO(pmalek) handle other types of CP ref
+	// TODO(pmalek) handle cross namespace refs
+	nnCP := types.NamespacedName{
+		Namespace: consumer.Namespace,
+		Name:      consumer.Spec.ControlPlaneRef.KonnectNamespacedRef.Name,
+	}
+	var cp konnectv1alpha1.KonnectControlPlane
+	if err := cl.Get(ctx, nnCP, &cp); err != nil {
+		return fmt.Errorf("failed to get KonnectControlPlane %s: for %T %s: %w",
+			nnCP, consumer, client.ObjectKeyFromObject(consumer), err,
+		)
+	}
+
+	if cp.Status.ID == "" {
+		return fmt.Errorf(
+			"can't update %T when referenced KonnectControlPlane %s does not have the Konnect ID",
+			consumer, nnCP,
+		)
+	}
+	if consumer.Status.Konnect == nil || consumer.Status.Konnect.ID == "" {
+		return fmt.Errorf(
+			"can't update %T %s when it does not have the Konnect ID",
+			consumer, client.ObjectKeyFromObject(consumer),
+		)
+	}
+
+	resp, err := sdk.Consumers.UpsertConsumer(ctx,
+		sdkkonnectgoops.UpsertConsumerRequest{
+			ControlPlaneID: cp.Status.ID,
+			ConsumerID:     consumer.GetKonnectStatus().GetKonnectID(),
+			Consumer:       kongConsumerToSDKConsumerInput(consumer),
+		},
+	)
+
+	// TODO: handle already exists
+	// Can't adopt it as it will cause conflicts between the controller
+	// that created that entity and already manages it, hm
+	if errWrapped := wrapErrIfKonnectOpFailed[configurationv1.KongConsumer](err, UpdateOp, consumer); errWrapped != nil {
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				KonnectEntityProgrammedConditionType,
+				metav1.ConditionFalse,
+				"FailedToCreate",
+				errWrapped.Error(),
+				consumer.GetGeneration(),
+			),
+			consumer,
+		)
+		return errWrapped
+	}
+
+	consumer.Status.Konnect.SetKonnectID(*resp.Consumer.ID)
+	consumer.Status.Konnect.SetControlPlaneID(cp.Status.ID)
+	k8sutils.SetCondition(
+		k8sutils.NewConditionWithGeneration(
+			KonnectEntityProgrammedConditionType,
+			metav1.ConditionTrue,
+			KonnectEntityProgrammedReason,
+			"",
+			consumer.GetGeneration(),
+		),
+		consumer,
+	)
+
+	return nil
+}
+
+func deleteConsumer(
+	ctx context.Context,
+	sdk *sdkkonnectgo.SDK,
+	consumer *configurationv1.KongConsumer,
+) error {
+	id := consumer.Status.Konnect.GetKonnectID()
+	if id == "" {
+		return fmt.Errorf("can't remove %T without a Konnect ID", consumer)
+	}
+
+	_, err := sdk.Consumers.DeleteConsumer(ctx, consumer.Status.Konnect.ControlPlaneID, id)
+	if errWrapped := wrapErrIfKonnectOpFailed[configurationv1.KongConsumer](err, DeleteOp, consumer); errWrapped != nil {
+		// Consumer delete operation returns an SDKError instead of a NotFoundError.
+		var sdkError *sdkerrors.SDKError
+		if errors.As(errWrapped, &sdkError) && sdkError.StatusCode == 404 {
```

Comment:
```suggestion
		var sdkError *sdkerrors.NotFoundError
		if errors.As(err, &sdkError) {
```

Reply:
As noted above in the comment, for some reason this returns `SDKError` not `NotFoundError` 🤷 

https://github.com/Kong/sdk-konnect-go/blob/080a82b1dbf650f4eccf59f4452d4489915e66ad/consumers.go#L966-L969

Reply:
Ah, sorry I missed the comment.